File: androidgeneratedcode/jcrystal/mobile/entities/Product.java
package jcrystal.mobile.entities;
import jcrystal.mobile.*;
import jcrystal.mobile.entities.enums.*;
import jcrystal.datetime.*;
import static jcrystal.JSONUtils.*;
public class Product implements ProductNormal,jcrystal.mobile.net.ISerializable{
	private String image;
	public String image(){return this.image;}
	public void image(String val){image = val;}
	private Long id;
	public Long id(){return this.id;}
	public void id(Long val){id = val;}
	private String name;
	public String name(){return this.name;}
	public void name(String val){name = val;}
	private long category;
	public long category(){return this.category;}
	public void category(long val){category = val;}
	private double price;
	public double price(){return this.price;}
	public void price(double val){price = val;}
	private double discount;
	public double discount(){return this.discount;}
	public void discount(double val){discount = val;}
	private double oldPrice;
	public double oldPrice(){return this.oldPrice;}
	public void oldPrice(double val){oldPrice = val;}
	private long color;
	public long color(){return this.color;}
	public void color(long val){color = val;}
	private long size;
	public long size(){return this.size;}
	public void size(long val){size = val;}
	public Product(){
	}
	protected Product(org.json.JSONObject json)throws org.json.JSONException{
		this.image = json.has("image")&&!json.isNull("image")?json.getString("image"):null;
		this.id = json.has("id")&&!json.isNull("id")?json.getLong("id"):null;
		this.name = json.has("name")&&!json.isNull("name")?json.getString("name"):null;
		this.category = json.has("category")&&!json.isNull("category")?Categories.fromId(json.getInt("category")):null;
		this.price = json.optDouble("price");
		this.discount = json.optDouble("discount");
		this.oldPrice = json.optDouble("oldPrice");
		this.color = json.has("color")&&!json.isNull("color")?Color.fromId(json.getInt("color")):null;
		this.size = json.has("size")&&!json.isNull("size")?Size.fromId(json.getInt("size")):null;
	}
	public Product(org.json.JSONObject json, JsonLevel level)throws org.json.JSONException{
		if(level != null){
			switch(level){
				case NORMAL: {
					this.id = json.has("id")&&!json.isNull("id")?json.getLong("id"):null;
					this.name = json.has("name")&&!json.isNull("name")?json.getString("name"):null;
					this.category = json.has("category")&&!json.isNull("category")?Categories.fromId(json.getInt("category")):null;
					this.price = json.optDouble("price");
					this.discount = json.optDouble("discount");
					this.oldPrice = json.optDouble("oldPrice");
					this.color = json.has("color")&&!json.isNull("color")?Color.fromId(json.getInt("color")):null;
					this.size = json.has("size")&&!json.isNull("size")?Size.fromId(json.getInt("size")):null;
				}
				break;
			}
		}
		else{
			this.image = json.has("image")&&!json.isNull("image")?json.getString("image"):null;
			this.id = json.has("id")&&!json.isNull("id")?json.getLong("id"):null;
			this.name = json.has("name")&&!json.isNull("name")?json.getString("name"):null;
			this.category = json.has("category")&&!json.isNull("category")?Categories.fromId(json.getInt("category")):null;
			this.price = json.optDouble("price");
			this.discount = json.optDouble("discount");
			this.oldPrice = json.optDouble("oldPrice");
			this.color = json.has("color")&&!json.isNull("color")?Color.fromId(json.getInt("color")):null;
			this.size = json.has("size")&&!json.isNull("size")?Size.fromId(json.getInt("size")):null;
		}
	}
	@Override public void toJson(java.io.PrintStream _pw){
		SerializerProduct.toJson(_pw, this);
	}
	public static class ListUtils{
		public static java.util.ArrayList<Product> listFromJson(org.json.JSONArray json)throws org.json.JSONException{
			java.util.ArrayList<Product> ret = new java.util.ArrayList<Product>(json.length());
			for(int e = 0, i = json.length(); e < i; e++){
				ret.add(new Product(json.getJSONObject(e)));
			}
			return ret;
		}
		public static java.util.ArrayList<ProductNormal> listFromJsonNormal(org.json.JSONArray json)throws org.json.JSONException{
			java.util.ArrayList<ProductNormal> ret = new java.util.ArrayList<ProductNormal>(json.length());
			for(int e = 0, i = json.length(); e < i; e++){
				ret.add(new Product(json.getJSONObject(e), JsonLevel.NORMAL));
			}
			return ret;
		}
	}
	public static class MapList{
		public static<T extends ProductNormal> java.util.Map<Integer, T> ByCategory(java.util.List<T> lista){
			java.util.Map<Integer, T> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.category() != null){
					ret.put(val.category().id, val);
				}
			}
			return ret;
		}
		public static<T extends ProductNormal> java.util.Map<Integer, T> ByColor(java.util.List<T> lista){
			java.util.Map<Integer, T> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.color() != null){
					ret.put(val.color().id, val);
				}
			}
			return ret;
		}
		public static<T extends ProductNormal> java.util.Map<Integer, T> BySize(java.util.List<T> lista){
			java.util.Map<Integer, T> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.size() != null){
					ret.put(val.size().id, val);
				}
			}
			return ret;
		}
	}
	public static class Group{
		public static<T extends ProductNormal> java.util.Map<Integer, java.util.List<T>> ByCategory(java.util.List<T> lista){
			java.util.Map<Integer, java.util.List<T>> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.category() != null){
					java.util.List<T> list = ret.get(val.category().id);
					if(list == null){
						ret.put(val.category().id, list = new java.util.ArrayList<T>());
					}
					list.add(val);
				}
			}
			return ret;
		}
		public static<T extends ProductNormal> java.util.Map<Integer, java.util.List<T>> ByColor(java.util.List<T> lista){
			java.util.Map<Integer, java.util.List<T>> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.color() != null){
					java.util.List<T> list = ret.get(val.color().id);
					if(list == null){
						ret.put(val.color().id, list = new java.util.ArrayList<T>());
					}
					list.add(val);
				}
			}
			return ret;
		}
		public static<T extends ProductNormal> java.util.Map<Integer, java.util.List<T>> BySize(java.util.List<T> lista){
			java.util.Map<Integer, java.util.List<T>> ret = new java.util.TreeMap<>();
			for(T val : lista){
				if(val.size() != null){
					java.util.List<T> list = ret.get(val.size().id);
					if(list == null){
						ret.put(val.size().id, list = new java.util.ArrayList<T>());
					}
					list.add(val);
				}
			}
			return ret;
		}
	}
}

File: iosgeneratedcode/jcrystal/mobile/entities/Product.swift
import Foundation
import jCrystaliOSLib
class Product: ProductNormal{
	fileprivate var _id : Int64?
	fileprivate var _name : String?
	fileprivate var _description : String?
	fileprivate var _category : Int64 = Int64()
	fileprivate var _price : Double = 0
	fileprivate var _discount : Double = 0
	fileprivate var _oldPrice : Double = 0
	fileprivate var _color : Int64 = Int64()
	fileprivate var _size : Int64 = Int64()
	fileprivate var _image : String?
	
	public func id()->Int64? {return self._id}
	public func id(val : Int64?){self._id = val}
	public func name()->String? {return self._name}
	public func name(val : String?){self._name = val}
	public func description()->String? {return self._description}
	public func description(val : String?){self._description = val}
	public func category()->Int64 {return self._category}
	public func category(val : Int64){self._category = val}
	public func price()->Double {return self._price}
	public func price(val : Double){self._price = val}
	public func discount()->Double {return self._discount}
	public func discount(val : Double){self._discount = val}
	public func oldPrice()->Double {return self._oldPrice}
	public func oldPrice(val : Double){self._oldPrice = val}
	public func color()->Int64 {return self._color}
	public func color(val : Int64){self._color = val}
	public func size()->Int64 {return self._size}
	public func size(val : Int64){self._size = val}
	public func image()->String? {return self._image}
	public func image(val : String?){self._image = val}
	init(){
	}
	class MapList{
		static func ByCategory(_ a : [ProductNormal]) -> Dictionary<Int, ProductNormal>{
			return Dictionary(a.map({($0.category()!.id, $0)}), uniquingKeysWith : { (first, _) in first })
		}
		static func ByColor(_ a : [ProductNormal]) -> Dictionary<Int, ProductNormal>{
			return Dictionary(a.map({($0.color()!.id, $0)}), uniquingKeysWith : { (first, _) in first })
		}
		static func BySize(_ a : [ProductNormal]) -> Dictionary<Int, ProductNormal>{
			return Dictionary(a.map({($0.size()!.id, $0)}), uniquingKeysWith : { (first, _) in first })
		}
	}
	class Group{
		static func ByCategory(_ a : [ProductNormal]) -> Dictionary<Int, [ProductNormal]>{
			return Dictionary(grouping: a, by: {$0.category()!.id})
		}
		static func ByColor(_ a : [ProductNormal]) -> Dictionary<Int, [ProductNormal]>{
			return Dictionary(grouping: a, by: {$0.color()!.id})
		}
		static func BySize(_ a : [ProductNormal]) -> Dictionary<Int, [ProductNormal]>{
			return Dictionary(grouping: a, by: {$0.size()!.id})
		}
	}
}
class SerializerProduct{
	public static func fromJson(_ json : [String: AnyObject]) -> Product{
		let ret = Product()
		if let numberVal = json["id"] as? NSNumber {
			ret._id = numberVal.int64Value
		}
		else{
			ret._id = Int64(json["id"] as? String ?? "0")
		}
		ret._name = json["name"] as? String
		ret._description = json["description"] as? String
		if let _category = json["category"] as? Int {
			ret._category = Categories.fromId(_category)
		}
		ret._price = json["price"] as? Double ?? 0.0
		ret._discount = json["discount"] as? Double ?? 0.0
		ret._oldPrice = json["oldPrice"] as? Double ?? 0.0
		if let _color = json["color"] as? Int {
			ret._color = Color.fromId(_color)
		}
		if let _size = json["size"] as? Int {
			ret._size = Size.fromId(_size)
		}
		ret._image = json["image"] as? String
		return ret
	}
	public static func listFromJson(_ json : [[String: AnyObject]]) -> [Product]{
		var ret = [Product]()
		for e in 0 ..< json.count{
			ret.append(fromJson(json[e]))
		}
		return ret
	}
	public static func toJson(_ _pw : OutputStream, objeto : Product){
		_pw.print("{");
		_pw.print("\"price\":", objeto._price);
		_pw.print(",\"discount\":", objeto._discount);
		_pw.print(",\"oldPrice\":", objeto._oldPrice);
		if let val3 = objeto._id {
			_pw.print(",\"id\":", val3);
		}
		if let val4 = objeto._name {
			_pw.print(",\"name\":", jsonQuote(val4));
		}
		if let val5 = objeto._description {
			_pw.print(",\"description\":", jsonQuote(val5));
		}
		if let val6 = objeto._category {
			_pw.print(",\"category\":", val6.id);
		}
		if let val7 = objeto._color {
			_pw.print(",\"color\":", val7.id);
		}
		if let val8 = objeto._size {
			_pw.print(",\"size\":", val8.id);
		}
		if let val9 = objeto._image {
			_pw.print(",\"image\":", jsonQuote(val9));
		}
		_pw.print("}");
	}
	public static func toJsonProduct(_ _pw : OutputStream, lista : [Product]){
		_pw.print("[");
		var p : Bool = false
		for valor in lista{
			if p {_pw.print(",");}
			SerializerProduct.toJson(_pw, objeto: valor);
			p = true;
		}
		_pw.print("]");
	}
	public static func toJsonObject(objeto : Product) -> [String:Any]{
		var __ret : [String:Any] = [String:Any]()
		if let val0 = objeto._id {
			__ret["id"] = val0;
		}
		if let val1 = objeto._name {
			__ret["name"] = val1;
		}
		if let val2 = objeto._description {
			__ret["description"] = val2;
		}
		if let val3 = objeto._category {
			__ret["category"] = val3.id;
		}
		__ret["price"] = objeto._price;
		__ret["discount"] = objeto._discount;
		__ret["oldPrice"] = objeto._oldPrice;
		if let val7 = objeto._color {
			__ret["color"] = val7.id;
		}
		if let val8 = objeto._size {
			__ret["size"] = val8.id;
		}
		if let val9 = objeto._image {
			__ret["image"] = val9;
		}
		return __ret;
	}
	public static func toJsonArray(_ lista:[Product])->[[String:Any]]{
		var __ret : [[String:Any]] = [[String:Any]]()
		for valor in lista{
			__ret.append(SerializerProduct.toJsonObject(objeto: valor))
		}
		return __ret;
	}
	public static func fromJsonNormal(_ json : [String: AnyObject]) -> Product{
		let ret = Product()
		if let numberVal = json["id"] as? NSNumber {
			ret._id = numberVal.int64Value
		}
		else{
			ret._id = Int64(json["id"] as? String ?? "0")
		}
		ret._name = json["name"] as? String
		ret._description = json["description"] as? String
		if let _category = json["category"] as? Int {
			ret._category = Categories.fromId(_category)
		}
		ret._price = json["price"] as? Double ?? 0.0
		ret._discount = json["discount"] as? Double ?? 0.0
		ret._oldPrice = json["oldPrice"] as? Double ?? 0.0
		if let _color = json["color"] as? Int {
			ret._color = Color.fromId(_color)
		}
		if let _size = json["size"] as? Int {
			ret._size = Size.fromId(_size)
		}
		ret._image = json["image"] as? String
		return ret
	}
	public static func listFromJsonNormal(_ json : [[String: AnyObject]]) -> [ProductNormal]{
		var ret = [ProductNormal]()
		for e in 0 ..< json.count{
			ret.append(fromJsonNormal(json[e]))
		}
		return ret
	}
	static let CREATOR = {(_ json: [String: AnyObject])->(Product) in 
		return SerializerProduct.fromJson(json)
	}
}
